### PR TITLE
Fix module

### PR DESCRIPTION
**Pull request by roberth**
and add unit tests

- [x] rebase

### CODE REVIEW

File: for-upstream/common.nix [hunk 1]
```diff
@@ -92,18 +100,25 @@ in
...
         }
         // cfg.extraOptions
       );
-
-      # TODO: expose only the (future) main directory as an option and derive
-      # all locations from finalConfig.
-      extraOptions.clusterJoinTokenPath = lib.mkDefault (cfg.secretsDirectory + "/cluster-join-token.key");
+      effectiveConfig =
+        let
+          effectiveConfig = defaults // cfg.fileConfig;
```
Comment (domenkozar) [inline for-upstream/common.nix]: Best to inline this
```suggestion
```

File: for-upstream/common.nix [hunk 2]
```diff
@@ -92,18 +100,25 @@ in
...
+          defaults = {
+            workDirectory = effectiveConfig.baseDirectory + "/work";
+            staticSecretsDirectory = effectiveConfig.baseDirectory + "/secrets";
+            clusterJoinTokenPath = effectiveConfig.staticSecretsDirectory + "/cluster-join-token.key";
+            binaryCachesPath = effectiveConfig.staticSecretsDirectory + "/binary-caches.json";
+          };
+        in
+          effectiveConfig;
```
Comment (domenkozar) [inline for-upstream/common.nix]:
```suggestion
          defaults // cfg.fileConfig;
```

File: for-upstream/common.nix (hunk 1 above)
Comment (domenkozar) [reply for-upstream/common.nix]: Ah nevermind, defaults depend on effectiveConfig...

Comment (roberth) [inline for-upstream/common.nix]:
```suggestion
          # recursively compute the effective configuration
          effectiveConfig = defaults // cfg.fileConfig;
```

**Comment by domenkozar** [comment]
I thought you changed the logic to figure out defaults for binary-caches.json in Haskell, so I'm not entirely sure what happened.

**Comment by roberth** [comment]
It's duplicated so that the key deployment module knows where to put the files.

**Pull request merged**